### ISSUE TITLE
Abstracting IO timeout settings

### ISSUE THREAD

**Issue by straight-shoota** [crystal-lang/crystal#10774]
`IO::Evented` defines `read_timeout` and `write_timeout` properties. They are not specific to the evented IO implementation. Any other implementation would use such timeouts, as well.
For win32's overlapped IO implementation, I copied the identical definitions from `IO::Evented`:

https://github.com/crystal-lang/crystal/blob/ef3dc9f0ca5e7ed06cd6d57adc6bd11073c21e8b/src/io/evented.cr#L18-L48

Besides, these methods are part of the public API (of `Socket` and `IO::FileDescriptor`), so they should really be identical across different IO implementations.

Thus, I propose to abstract them to a dedicated module `IO::Timeout` which would define those two ivars and their getters and setters. `Socket` and `IO::FileDescriptor` should directly include this module.

Additionally, there are very similar methods in `HTTP::Client`. The only difference is that the values are stored as `Float` instead of `Time::Span`. But that could easily be adjusted to use the same module. The external API is already identical.

https://github.com/crystal-lang/crystal/blob/ef3dc9f0ca5e7ed06cd6d57adc6bd11073c21e8b/src/http/client.cr#L269-L313

Note: `HTTP::Client` stores these values instead of directly delegating to the underlying socket, because a client can use different socket instances under the hood (when it gets closed, it can re-open a new one for example).

There are also related `connect_timeout` and `dns_timeout` properties on `HTTP::Client` which don't exist on `Socket`. They're only methods parameters there.

For the record, `IO::Timeout` was the previous name of `IO::TimeoutError`. This could potentially be a reason for confusion. But the error type was renamed in 0.34.0 (#8885) without a deprecation period and I assume that's far enough away that it shouldn't prevent the name being repurposed.

**Comment by yxhuvud** [comment]
I agree that it should not belong to `Evented` but I am not certain there is a need of an abstraction. It is just two accessors with a default value after all.

**Comment by straight-shoota** [comment]
With the overloads for different argument types, it still totals to six different method definitions. IMO at that point an abstraction could be valuable.

**Comment by straight-shoota** [comment]
An additional argument: These methods are part of the public API. They should be portable and documented in a single place.

#10766 even requires an implementation that is independent of the async IO backend.

**Comment by Blacksmoke16** [comment]
With https://github.com/crystal-lang/crystal/issues/14368 being resolved now, is there still a desire to add this `IO::Timeout` module/interface? It would pair well with some of the conversation in https://github.com/crystal-lang/crystal/issues/14799.
